### ISSUE TITLE
Spurious SC2031, shellcheck thinks variable is used by associative array index with similar name

### ISSUE THREAD

**Issue by wolfgang42** [koalaman/shellcheck#2390]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2030, SC2031
- My shellcheck version (`shellcheck --version` or "online"): online
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash
declare -A foo=()

(bar=8)

echo "${foo[bar-baz]}"
```

(this also happens with just `${foo[bar]}` but the above version is even more confusing)

#### Here's what shellcheck currently says:
```
Line 4:
(bar=8)
 ^-- SC2030 (info): Modification of bar is local (to subshell caused by (..) group).
 
Line 6:
echo "${foo[bar-baz]}"
        ^-- SC2031 (info): bar was modified in a subshell. That change might be lost.
```

#### Here's what I wanted or expected to see:
```
Line 4:
(bar=8)
 ^-- SC2034 (warning): bar appears unused. Verify use (or export if used externally).
```
(or no warning at all if the variable is actually used, but only inside the subshell)
